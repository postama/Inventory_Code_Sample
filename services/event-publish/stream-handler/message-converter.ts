import { GetRecordsOutput } from 'aws-sdk/clients/dynamodbstreams';
import { Converter } from 'aws-sdk/clients/dynamodb';
import { PublishInput } from 'aws-sdk/clients/sns';

export function convertDynamoToSNSMessage(event: GetRecordsOutput): PublishInput[] {
    const records = event.Records.map(r => Converter.unmarshall(r.dynamodb.NewImage));
    const messages = records.map(r => ({
        Message: JSON.stringify(r),
        Subject: r.type,
        TopicArn: process.env.snsArn,
        MessageAttributes: {
            Type: {
                DataType: 'String',
                StringValue: r.type
            }
        }
    }));
    return messages;
}