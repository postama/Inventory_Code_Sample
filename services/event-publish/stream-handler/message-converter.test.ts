import { convertDynamoToSNSMessage } from './message-converter';

describe('message converter', () => {
    beforeAll(() => process.env.snsArn = 'test');

    test('should create a sns event for one record', () => {
        const result = convertDynamoToSNSMessage(createMessages([{S: 'test'}]));
        expect(result).toMatchSnapshot();
    });

    test('should create an sns event for each record (2)', () => {
        const result = convertDynamoToSNSMessage(createMessages([{S: 'test'}, {S: 'test2'}]));
        expect(result).toMatchSnapshot();
    });
});

function createMessages(items) {
    return {
        Records: items.map(i => ({
            eventID: 'test',
            eventName: 'INSERT',
            eventVersion: '1.1',
            eventSource: 'aws:dynamodb',
            awsRegion: 'test',
            dynamodb: {
                ApproximateCreationDateTime: new Date('2019-03-21T21:36:00.000Z'),
                Keys: {
                    customerId: { S: 'test' },
                    commandId: { S: 'test' }
                },
                NewImage: {
                    test: i
                },
                SequenceNumber: 'test',
                SizeBytes: 100,
                StreamViewType: 'NEW_IMAGE'
            }
        })),
        NextShardIterator: 'test'
    };
}
