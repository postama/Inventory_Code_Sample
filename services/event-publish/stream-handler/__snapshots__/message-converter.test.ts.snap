// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`message converter should create a sns event for one record 1`] = `
Array [
  Object {
    "Message": "{\\"test\\":\\"test\\"}",
    "MessageAttributes": Object {
      "Type": Object {
        "DataType": "String",
        "StringValue": undefined,
      },
    },
    "Subject": undefined,
    "TopicArn": "test",
  },
]
`;

exports[`message converter should create an sns event for each record (2) 1`] = `
Array [
  Object {
    "Message": "{\\"test\\":\\"test\\"}",
    "MessageAttributes": Object {
      "Type": Object {
        "DataType": "String",
        "StringValue": undefined,
      },
    },
    "Subject": undefined,
    "TopicArn": "test",
  },
  Object {
    "Message": "{\\"test\\":\\"test2\\"}",
    "MessageAttributes": Object {
      "Type": Object {
        "DataType": "String",
        "StringValue": undefined,
      },
    },
    "Subject": undefined,
    "TopicArn": "test",
  },
]
`;
